In [3]:
# importing the libraries
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error, r2_score
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [5]:
# Loading the '50_Startups.csv' dataset into a DataFrame
data=pd.read_csv('50_Startups.csv')
data.head()

R&D Spend  Administration  Marketing Spend       State     Profit
0  165349.20       136897.80        471784.10    New York  192261.83
1  162597.70       151377.59        443898.53  California  191792.06
2  153441.51       101145.55        407934.54     Florida  191050.39
3  144372.41       118671.85        383199.62    New York  182901.99
4  142107.34        91391.77        366168.42     Florida  166187.94

In [7]:
X = data.drop(columns=['Profit'])
y = data['Profit']

In [8]:
# Displaying the shape (dimensions) of the independent variables (X) and dependent variable (y)
print("Shape of X:", X.shape)
print("Shape of y:", y.shape)

Shape of X: (50, 4)
Shape of y: (50,)


In [9]:
X = X.values
y = y.values

In [10]:
#training (80%) and testing (20%) data sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [11]:
# Displaying the shape (dimensions) of the training data (X_train) and training target (y_train)
print("Shape of X:", X_train.shape)
print("Shape of y:", y_train.shape)

Shape of X: (40, 4)
Shape of y: (40,)


In [29]:
# Initialize and train the multiple linear regression model.
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set.
y_pred = model.predict(X_test)

# Evaluate the model.
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Multiple Linear Regression")
print(f"Mean Squared Error: {mse:.4f}")
print(f"R-squared: {r2:.4f}")

Multiple Linear Regression
Mean Squared Error: 80929465.4910
R-squared: 0.9001


In [30]:
print(model_multiple.intercept_)

54080.71684611373


In [31]:
print(round(model_multiple.intercept_,2))

54080.72


In [32]:
#print the data type of model_multiple
type(model_multiple)

sklearn.linear_model._base.LinearRegression

In [33]:
model_multiple.predict(X[0:5])

array([192405.36723938, 188372.32164914, 183282.8185662 , 174012.86095334,
       173529.22121764])

In [34]:
import pandas as pd
# Display the first few predictions alongside the actual values
predictions = pd.DataFrame({'Actual': y_test, 'Predicted': y_pred_multiple})
print(predictions.head())

       Actual      Predicted
13  134307.35  126720.661507
39   81005.76   84909.089619
30   99937.59   98890.318549
45   64926.08   46479.312402
17  125370.37  129113.183188


In [35]:
import pickle

In [39]:
# save: 'with' is package deals with file handeling, wb- write
with open('model.pkl','wb') as f:
    pickle.dump(model_multiple,f)
# load: rb-read
with open('model.pkl', 'rb') as f:
    clf2 = pickle.load(f)


In [41]:
# Using the trained model (clf2) to predict the target values for the first 5 instances of the dataset X
clf2.predict(X[0:5])

array([192405.36723938, 188372.32164914, 183282.8185662 , 174012.86095334,
       173529.22121764])

Analysis of Multiple Linear Regression:

The dataset consists of 50 instances and 4 features. After splitting, the
training set has 40 instances, and the test set has 10 instances.
The Multiple Linear Regression model has an R-squared value of 0.9001, indicating a strong fit to the data and mean squared error of 80929465.49.
The intercept of the model is approximately 54080.72.
The model makes predictions on the test set, with values close to the actual "Profit" values.